### PR TITLE
UCP/TAG/SEND: added eager/rndv flags to tag_send op

### PR DESCRIPTION
**Pull request by hoopoepg**
- added op-flags support to tag_send_[sync]_nbx functions
- added flags eager/rndv flags to force used protocol

backport from https://github.com/yosefe/ucx/pull/172

### CODE REVIEW

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@hoopoepg please fix the test failure

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -2607,6 +2613,8 @@ static int parse_args(int argc, char **argv, options_t *test_opts)
             std::cout << "  -D                          Enable debugging mode for IO operation timeouts" << std::endl;
             std::cout << "  -H                          Use human-readable timestamps" << std::endl;
             std::cout << "  -P <interval>               Set report printing interval"  << std::endl;
+            std::cout << "  -R <rndv-thresh>            Rendezvous threshold used to force eager" << std::endl;
```
Comment (evgeny-leksikov) [inline test/apps/iodemo/io_demo.cc]: ~wired~ *weird* description
1) maybe "Rendezvous threshold used to switch to rendezvous protocol"?
2) what is defalt value? I think if not set it should be selected by UCX

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc]: default value is -2 which means "auto" - selected by UCX

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: right, pls mention default behavior in description

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc]: ok, updated description

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -2614,7 +2614,10 @@ static int parse_args(int argc, char **argv, options_t *test_opts)
             std::cout << "  -H                          Use human-readable timestamps" << std::endl;
             std::cout << "  -P <interval>               Set report printing interval"  << std::endl;
             std::cout << "  -R <rndv-thresh>            Rendezvous threshold used to force eager" << std::endl;
-            std::cout << "                              or rendezvous protocol" << std::endl;
+            std::cout << "                              or rendezvous protocol:" << std::endl;
+            std::cout << "                                -2 - threshold is selected by UCX [default]" << std::endl;
+            std::cout << "                                -1 - force use eager protocol" << std::endl;
```
Comment (evgeny-leksikov) [inline test/apps/iodemo/io_demo.cc]: IMHO better don't mention magic values:
```
 - if not set - threshold is selected by UCX
 - to force always use eager protocol - set value larger than max data can be sent (see -d option)
```

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc]: ok, updated

**Review by evgeny-leksikov — APPROVED** [review]

**Comment by hoopoepg** [comment]
@yosefe are you ok with such wording?

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -2607,6 +2613,11 @@ static int parse_args(int argc, char **argv, options_t *test_opts)
             std::cout << "  -D                          Enable debugging mode for IO operation timeouts" << std::endl;
             std::cout << "  -H                          Use human-readable timestamps" << std::endl;
             std::cout << "  -P <interval>               Set report printing interval"  << std::endl;
+            std::cout << "  -R <rndv-thresh>            Rendezvous threshold used to force eager" << std::endl;
+            std::cout << "                              or rendezvous protocol:" << std::endl;
+            std::cout << "                                - if not set - threshold is selected by UCX" << std::endl;
+            std::cout << "                                - to force always use eager protocol - set value larger" << std::endl;
+            std::cout << "                                  than max data can be sent (see -d option)" << std::endl;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]:
```
- R <threshold>   Always use rendezvous protocol for messages larger than 
                  this size, and eager protocol for messages lower than this size.
                  If not set, the threshold is selected automatically by UCX

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc]: updated

Comment (brminich) [reply test/apps/iodemo/io_demo.cc]: then what protocol will be used for messages of exactly the `size`? Maybe
`Always use rendezvous protocol for messages starting from this size..`?

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc]: no problemo, re-worded

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Comment by yosefe** [comment]
@hoopoepg pls squash and rebase

**Comment by hoopoepg** [comment]
@yosefe squashed & rebased

**Comment by yosefe** [comment]
@hoopoepg need to rebase again :)

**Comment by hoopoepg** [comment]
yep, rebased again

**Pull request merged**